Flickr/Sources/FlickrAppVCs/SearchScreenVC.swift

```diff
@@ -171,41 +171,33 @@ class SearchScreenVC: UIViewController {
     }
 
     func fetchPhotos(searchString: String) {
+        if task?.state == .running {
+            task?.cancel()
+        }
         var photosArray: [URL] = []
         imageTitles = []
         imageIDs = []
         imagesLoaded = false
         photoSections[0].emptyPhotosArray()
-        guard let url = returnSearchUrl(searchString: searchString) else {
-            return
-        }
-        task = URLSession.shared.dataTask(with: url, completionHandler: { data, _, error in
-            guard let data = data, error == nil else {
-                return
-            }
-            var result: Photos?
-            do {
-                result = try JSONDecoder().decode(Photos.self, from: data)
-            } catch {
-                return
-            }
-            guard let result = result else {
-                return
-            }
-            photosArray = self.constructIndividualUrls(result)
-            self.photos = photosArray
-            DispatchQueue.main.async { [self] in
-                if let dataSource = dataSource {
-                    Observable.just(photoSections)
-                        .bind(to: collectionView.rx.items(dataSource: dataSource))
-                        .disposed(by: disposeBag)
+        ApiCallHandler.shared.fetchRequest(searchString: searchString)
+            .observe(on: MainScheduler.instance)
+            .subscribe { result in
+                switch result {
+                case let .success(response):
+                    photosArray = self.constructIndividualUrls(response)
+                    self.photos = photosArray
+                    if let dataSource = self.dataSource {
+                        Observable.just(self.photoSections)
+                            .bind(to: self.collectionView.rx.items(dataSource: dataSource))
+                            .disposed(by: self.disposeBag)
```

Comment:
You are still creating another Observable event every time new search results are received. This will essentially bind the datasource to new event every time a new search result is received and where as the older bind will never be released. What we essentially want is to bind the datasource to the fetch request observable.